### ISSUE TITLE
@AfterMapping methods with non unique parameters

### ISSUE THREAD

**Issue by lazydays79** [mapstruct/mapstruct#1457]
Hi!

First of all: we are using MapStruct now for almost 2 years in a big project and it saved us a lot of time, headache and coding. So for this I want to give you a big thanks and thumbs up! :grin: 

But all of the sudden, MapStruct cost us a lot of analysis, research, nerves and coke... but I think we found some issue. Ok, not a real issue, it's more like an "undefined behaviour" which should somehow be taken care of.

Let's assume we have a mapping method (defined in an abstract class) includung an @AfterMapping method. The method signatures both of the mapping and the after mapping method include two "context parameters", both of type String.

I must admit, that the bug that now occurs is made by us: we forgot to add the @Context annotation to the String parameters. If we would have done it right, MapStruct would give up throwing an error:

_The types of @Context parameters must be unique._

But we missed to add the annotation, so MapStruct generated a mapper which is calling the AfterMapping method wrong: using the first String parameter twice.

So, is this a bug in MapStruct? No, not really. It's just undefined behaviour. No question about that.

Could MapStruct have saved us some hours, if it would complain about this? Yes, I guess so. ;-)

So, I suggest to either let MapStruct throw a compilation warning or even error, when @Context is missing. Or it should complain about non unique parameters as it would with @Context annotated parameters.

To make it more clear what I'm blathering about, I created a simple test project with might help you to understand what I mean:

[https://github.com/lazydays79/mapstructissue2](https://github.com/lazydays79/mapstructissue2)

Feel free to clone, compile, look at the generated mapper and enjoy the failing test. :laughing: 

If you won't fix this, I will totally accept it. But I guess it would be a lot more helpfull (and time saving) to let MapStruct fail on this undefined behaviour than to let it generale a wrong Mapper implementation in silence.

Best regards,
Michael

**Comment by sjaakd** [comment]
HI Michael,

Thanks for your extensive analysis. I'll have a look soon if we can generate  a  warning in this case.

Best regards,
SJaak

**Comment by filiphr** [comment]
I have to say reading such an issue is a joy to read. It's always nice to hear such nice words from users and we are glad that MapStruct has saved you time. Will try to do something so such pesky issues don't lose more time to you or other users.

**Comment by RaymondDubovik** [comment]
I had the same issue when I tried to upgrade from mapstruct 1.1.0.Final to mapstruct 1.2.0.Final soon after it's release.
All of the String parameters in @AfterMapping were receiving the value of the first string parameter just in OPs case, which were caught by the unit tests.
Unfortunately at that moment due to time constraints I did not investigate it deeper and simply reverted back to 1.1.0 which was generating code exactly as expected.

Around half a year later, I decided to upgrade to 1.2.0.Final.
During this time, the "offending mapper's" code has mutated and looks something like this:
`
@Mappings({
        @Mapping(target = "requestId", source = "request.id"),
        @Mapping(target = "s1", source = "s1"),
        @Mapping(target = "s2", source = "s2"),
        @Mapping(target = "s3", source = "s3"),
        @Mapping(target = "s4", source = "s4"),
        // [.. other mappings ..]
})
public abstract TargetDto toDto(Request request, Bank bank, List<OpeningHours> openingHours, String s1, String s2, String s3, String s4, String s5, String s6);
	
@AfterMapping
protected void toDto(Request request, Bank bank, List<OpeningHours> openingHours, String s1, String s2, String s3, String s4, String s5, String s6, @MappingTarget TargetDto dto) {
    // [.. using request, bank, openingHours, s5, s6 to update some fields in dto ..]
}
`

And I have started receiving the following exeption when upgrading to 1.2.0.Final and 1.3.0.Beta1:
	
`java: Internal error in the mapping processor: java.lang.OutOfMemoryError: Java heap space [..]`
	
Without prior knowledge, that multiple Strings in @Aftermapping method's signature were not working correctly, I assume it would require lots of time to debug.
	
I hope noone gets the same exception as me, but in case you do, I hope my post will help you identify your problem quicker.
P.S. Write unit tests.

**Comment by sjaakd** [comment]
@RaymondDubovik  thanks for your additions.. I did not get the time yet to find a solution here. All help is welcome.

Wrt. to Unit test: we do have them 😄 .. Honestly, a lot of effort goes into them. But they do not cover all scenarios.

**Comment by chris922** [comment]
I took a look into this and my first idea was to check the parameter names for the ambiguous parameters.. unfortunately this is not so easy so what seems to be possible with just a few smaller changes is to check that all parameter names must match as soon as there are ambiguous parameters.

An example with both mentioned cases:

```java
abstract Target map(Source source, String stringA, String stringB);

// only names of ambiguous parameters are matching:
@AfterMapping
void afterMap1(@MappingTarget Target target, Source notNamedSource, String stringA, String stringB) { ... }

// all parameter names are matching:
@AfterMapping
void afterMap2(@MappingTarget Target target, Source source, String stringA, String stringB) { ... }
```

@RaymondDubovik: You mentioned in 1.1.0 you didn't had this issue. Do you (or maybe @sjaakd ?) know if it was also required to have all parameter names matching (`afterMap2`) in such an ambiguous case or was it just required to have the names of the ambiguous parameters matching like `afterMap1`?


@sjaakd: If it is fine to presume that all parameter names must match as soon as we've got such an ambiguous case I can take over and submit a PR if you like.

**Comment by sjaakd** [comment]
> @sjaakd: If it is fine to presume that all parameter names must match as soon as we've got such an ambiguous case I can take over and submit a PR if you like.

That might be an approach... Drawback: you have to know and its prone to error (e.g. refactoring).. Anyway. We use this mechanism on other places as well (you can/ need to mention the parameter name in the nested mapping -especially with multi sources - to avoid conflicts). 

So: what would you do if there's an ambiguity like this: 

```java
abstract Target map(Source source, String stringA, String stringB);

// only names of ambiguous parameters are matching:
@AfterMapping
void afterMap1(@MappingTarget Target source, Source notNamedSource, String stringA, String stringB) { ... }
```
are you only considering conflicting types (so, the Strings stringA and String stringB) and ignore the `Target` marked `source`?

**Comment by chris922** [comment]
>are you only considering conflicting types (so, the Strings stringA and String stringB) and ignore the Target marked source?

No, I planned this but this looks much more complicated to realize. Thus I would expect to have all parameter names matching as soon as you've got conflicting types. In case someone is already using conflicting/ambiguous types it should already be buggy, thus requiring to have all parameter names matching is imho fine. If this is not the case a warning would occur, so even then the user should be informed that something is most likely not working as expected.

This line here is more or less the entry point:
https://github.com/mapstruct/mapstruct/blob/master/processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/TypeSelector.java#L121
`parameterAssignments` is one possible permutation of the parameter arguments. Unfortunately only the `Type`s will be compared, thus the parameter name isn't available anymore. Additionally within the `Method#matches` method that knows how to check the parameters it is also not known if there is more than one matching permutation.... (hopefully that's halfway understandable 😆 )

If you like I can quickly rework my draft and supply a PR including tests.

**Comment by sjaakd** [comment]
@chris922 I'll have a look at your your ref asap. There's one concern: if the `@BeforeMapping`, `@AfterMapping` or `@ObjectFactory` is in another jar, you don't have the parameter names (but names like `arg1`, `arg2`..).. That's tricky to explain.

We could do a staged approach:
* resolve the `java.lang.OutOfMemoryError`
* match on parameter names as you suggest (in case of conflict) 
* next step (see below)

So.. Perhaps its better to use an annotation. Now, `@Context` as suggested above is not usable, since there can be multiple sources as well. So then we would have to invent one. Like `@Arg(name="stringA")`.

**Comment by sjaakd** [comment]
@chris922 please go ahead.. You're working on the right spot.

For the `@Arg(name="stringA")`, a similar thing is required when using constructor initialisation of target beans.. So we could wait, or do this separately.. WDYT?

**Comment by sjaakd** [comment]
@chris922 are you still working on this one? Perhpas we should allign the third bullit (next steps) with a solution for #73 .. Its a similar problem.

**Comment by chris922** [comment]
Haven't worked on this again since my last comment. I will try to catch up and see what I can supply here. Are you able to assign this to me so that I can better keep track? This issue was hidden by a lot of other topics that came up here :smile:

**Comment by Kaoxyd** [comment]
Any news about this ? I just encounter the same issue (and i'm glad to see that a advances analysis has been provided yet :)

Version : 1.3.0.Final

**Comment by sjaakd** [comment]
@Kaoxyd .. sorry no additional progress yet..

**Comment by chris922** [comment]
Fixed by PR #1782
